File: Python/53_MaximumSubarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        endPoint = nums[0]
        res = nums[0]

        for i in range(1,len(nums)):
            endPoint = max(endPoint+nums[i], nums[i])
            res = max(res, endPoint)

        return res

#kadane's algorithm
#nums       -2,1-3,4
#endPoint   -2,1,1,4
#res        -2,1,1,4

File: C++/53_MaximumSubarray.cpp
using namespace std;
#include <vector>

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int sum = 0;
        int total = nums[0];

        for (auto num:nums){
            sum = max(num, num + sum);
            total = max(sum, total);
        }

        return total;
    }
};

/*
 curT = max(num, curT+nums)
 maxT = max(maxT, curT)
*/
